port/cpl_float.cpp

```diff
@@ -306,3 +309,129 @@ float CPLConvertHalfToFloat(GUInt16 nHalf)
     std::memcpy(&fFloat32, &nFloat32, sizeof fFloat32);
     return fFloat32;
 }
+
+namespace
+{
+
+template <typename T> struct Fraction
+{
+    using value_type = T;
+
+    T num;
+    T denom;
+};
+
+/** Approximate a floating point number as a fraction, using the method describe
+ * in Richards, Ian (1981). Continued Fractions Without Tears. Mathematics
+ * Magazine, Vol. 54, No. 4. https://doi.org/10.2307/2689627
+ *
+ * If the fraction cannot be approximated within the specified error tolerance
+ * in a certain amount of iterations, a warning will be raised and  std::nullopt
+ * will be returned.
+ *
+ * @param x the number to approximate as a fraction
+ * @param err the maximum allowable absolute error in the approximation
+ *
+ * @return the approximated value, or std::nullopt
+ *
+*/
+std::optional<Fraction<std::uint64_t>> FloatToFraction(double x, double err)
+{
+    using inttype = std::uint64_t;
+    int nMaxIter = 1000000;
+
+    double sign = std::signbit(x) ? -1 : 1;
+
+    double g(std::abs(x));
+    inttype a(0);
+    inttype b(1);
+    inttype c(1);
+    inttype d(0);
+
+    Fraction<std::uint64_t> ret;
+
+    for (int i = 0; i < nMaxIter; i++)
+    {
+        inttype s = static_cast<inttype>(std::floor(g));
+        ret.num = a + s * c;
+        ret.denom = b + s * d;
+
+        a = c;
+        b = d;
+        c = ret.num;
+        d = ret.denom;
+        g = 1.0 / (g - static_cast<double>(s));
+
+        double approx = sign * static_cast<double>(ret.num) /
+                        static_cast<double>(ret.denom);
+
+        if (std::abs(approx - x) < err)
+        {
+            return ret;
+        }
+    }
+
+    CPLError(CE_Warning, CPLE_AppDefined,
+             "Failed to approximate %g as a fraction with error < %g in %d "
+             "iterations",
+             x, err, nMaxIter);
+    return std::nullopt;
+}
+}  // namespace
+
+/** Return the largest value by which two input values can be
+ *  divided, with the result being an integer. If no suitable
+ *  value can be found, zero will be returned.
+ */
+double CPLGreatestCommonDivisor(double a, double b)
```

Comment:
It could make sense to have a autotest/cpp/ test for CPLGreatestCommonDivisor()

Reply:
added